include/ode_solvers/leapfrog.hpp

```diff
@@ -71,31 +124,52 @@ struct LeapfrogODESolver {
 
       // x <- x + eta v'
       Point y = xs[v_index];
-      y = (eta) * y;
 
       if (Ks[x_index] == NULL) {
-        xs[x_index] = xs_prev[x_index] + y;
+        xs[x_index] = xs_prev[x_index] + eta*y;
       }
       else {
         // Find intersection (assuming a line trajectory) between x and y
-        do {
-
-          pbpair = Ks[x_index]->line_positive_intersect(xs_prev[x_index], y, Ar, Av);
-
-          if (pbpair.first >= 0 && pbpair.first <= 1) {
+        bool step_completed = false;
+        NT T = eta;
+        if (k == 0 && !accepted) {
+          Ar[x_index] = Ks[x_index]->get_mat() * xs_prev[x_index].getCoefficients();
```

Comment:
variable `k` corresponds to the walk length? If yes, then why `k==0` in this condition? If (k!=0 and !accepted) holds then the initialization in line 136 is again necessary, right?

Reply:
We have the following truth table:

*  `k == 0` and `accepted` that means that we can use the pre-cached data from the previous call to step() to continue the iteration as part of a "lengthier integration", so the initialization step is not needed. 
* `k != 0` and `accepted` we continue integration since this is called through the `steps()` method (see below).
* `k == 0` and `!accepted` the initialization must run since the previous proposal has not been accepted and we are at the first call of `steps()`.
* `k != 0` and `!accepted`  we have `k != 0` (we need `k` to run the `steps()` command where we use `step(i, accepted)`, discarding initialization at later iteration of the same call to `steps()` (plural). So we don't need to initialize. 

So the if clause is correct . Moreover, for the reinitialization, what do mean by Line 136? Line 136 is the first if. Do you mean another Line?

Reply:
Yes @papachristoumarios your answer is quite clear. Thank you. 
By `line 136` I mean the initialization `Ar[x_index] = Ks[x_index]->get_mat() * xs_prev[x_index].getCoefficients();`, which is necessary in the first step only when the pervious state was rejected.